gennet-module/gennet.py

```diff
@@ -221,51 +265,104 @@ def generate_and_write_files(dirname, num_topics, num_subnets, G):
+            with open(cfile, 'r') as f:  # Load config file
+                conf = json.load(f)
+                if "gennet" not in conf:
+                    print(f"Gennet configuration not found in {cfile}. Skipping topology generation.")
+                    sys.exit(1)
+                # TODO : type-check and sanity-check the config.json
+                #print(conf)
+        except Exception as ex:
+            raise typer.BadParameter(str(ex))
+    return cfile
+
+
+# methods to merge cli values, config.json and default values
+def test_and_set_int(cli_val, file_val, conf, module="gennet"):
+    if cli_val != INT_NONE:
+        return cli_val
+    if cli_val == INT_NONE and module in conf and file_val in conf[module]:
+        return conf[module][file_val]
+    return defaults[file_val]
+
+
+def test_and_set_str(cli_val, file_val, conf, module="gennet"):
+    if cli_val != STR_NONE:
+        return cli_val
+    if cli_val == STR_NONE and module in conf and file_val in conf[module]:
+        return conf[module][file_val]
+    return defaults[file_val]
+
+
+def main(benchmark : bool = False,
+         output_dir: str = STR_NONE,
+         prng_seed: int = INT_NONE,
+         num_nodes: int = INT_NONE,
+         num_topics: int = INT_NONE,
          network_type: networkType = networkType.NEWMANWATTSSTROGATZ.value,
-         num_subnets: int = typer.Option(-1, callback=_num_subnets_callback),
-         num_partitions: int = typer.Option(1, callback=_num_partitions_callback),
-         config_file: str = typer.Option("", callback=conf_callback, is_eager=True)):
+         num_partitions: int = typer.Option(INT_NONE, callback=_num_partitions_callback),
+         num_subnets: int = typer.Option(INT_NONE, callback=_num_subnets_callback),
+         config_file: str = typer.Option(STR_NONE, callback=conf_callback, is_eager=True)):
+
+    # Benchmarking: record start time and start tracing mallocs
+    if benchmark :
+        start = time.time()
+        tracemalloc.start()
+
+
+    conf = {}
+    if config_file != "" :
+        with open(config_file, 'r') as f:  # Load config file
+            conf = json.load(f)
+        #print(conf)
+
+    # set the random seed : networkx uses numpy.random as well
+    seed = test_and_set_int(prng_seed, "prng_seed", conf, "general")
+    print("Setting the random seed to", seed)
+    random.seed(seed)
+    np.random.seed(seed)
+   
+    # Extract the node type distribution from config.json or defaults
+    # no cli for node type distribution
+    if "gennet" in conf  and "node_type_distribution" in conf ["gennet"]:
+        node_type_distribution = conf["gennet"]["node_type_distribution"]
+    else:
+        node_type_distribution = defaults["node_type_distribution"]
+
+    # merge the cli options and the config.json options
+    # TODO : pack the fields in a class/'struct'/tuple
+    # TODO : use inspect and local() to do this iteratively
+    output_dir      =   test_and_set_str(output_dir, "output_dir", conf)
```

Comment:
Uhm, I don't understand why you did this "test_and_set". This functionality was already implemented but you deleted it. In this way, you will need to set up manually every new variable, and the way it was before, just with two lines, would map automatically the variables of the file with the arguments from the CLI, being these last ones with the higher priority.

Did you find any bug or something? Because I tested it and it was working perfectly.

Reply:
Well, the typer merge was not working: whatever the value set in config.json (via --config-file), gennet was generating the network only for the default values. Now gennet correctly prioritises the inputs in the following order: the cli, then config.json, then, if neither are set, the default values.

Reply:
I tested it, and tested it again before adding this comment, and it was working fince.

Were you using the good config.file? If you were using docker, did you rebuild the image?

Reply:
So, I tried this again, and having as a default:
```
         output_dir: str = "topology_generated",
         num_nodes: int = 4,
         num_topics: int = 1,
         network_type: networkType = networkType.NEWMANWATTSSTROGATZ.value,
         num_subnets: int = typer.Option(-1, callback=_num_subnets_callback),
         num_partitions: int = typer.Option(1, callback=_num_partitions_callback),
         config_file: str = typer.Option("", callback=conf_callback, is_eager=True))
```

Having this in my `config.json`:
```
  "gennet": {
    "num_nodes": 5,
    "num_topics": 7,
    "network_type": "scalefree",
    "num_subnets": 3
  }
```

And adding this in CLI `--config-file config.json --num-nodes 6` I am expecting a topology of 6 nodes (since it is given by CLI thus it has preference), with 7 topics and 3 subnets (given by config.file), and it is the restult I am having.

Basically how my code works is that the callback of `config_file` will be called first because of `is_eager` parameter. Then it will update the default values from typer with those values from the `config.json` file. So we have default -> json. Then, the rest of the parameters will be read as usual by CLI, so they will be overriten, having default -> json -> CLI, which is the behaviour we are looking for.

If this does not work for you, can you elaborate a step by step case on where you do not have the same behaviour?

Reply:
the call back part and the merge part is clear. the issue for me is the flow of values from the one call back context object to any (and all) of the input params. in any case, let me have a look at later today and see what's the issue, and fix the code accordingly.